core/src/main/java/org/apache/calcite/sql/SqlIntervalQualifier.java

```diff
@@ -1131,6 +1149,111 @@ private int[] evaluateIntervalLiteralAsSecond(
+   * value is illegal
+   */
+  private int[] evaluateIntervalLiteralAsMillisecond(
+      RelDataTypeSystem typeSystem, int sign,
+      String value,
+      String originalValue,
+      SqlParserPos pos) {
+    BigDecimal millisecond;
+
+    // validate as MINUTE(startPrecision), e.g. 'MM'
+    String intervalPattern = "(\\d+)";
+
+    Matcher m = Pattern.compile(intervalPattern).matcher(value);
+    if (m.matches()) {
+      // Break out  field values
+      try {
+        millisecond = parseField(m, 1);
+      } catch (NumberFormatException e) {
+        throw invalidValueException(pos, originalValue);
+      }
+
+      // Validate individual fields
+      checkLeadFieldInRange(typeSystem, sign, millisecond, TimeUnit.MILLISECOND, pos);
+
+      // package values up for return
+      return fillIntervalValueArray(sign, millisecond, ZERO, ZERO);
+    } else {
+      throw invalidValueException(pos, originalValue);
+    }
+  }
+
+  /**
+   * Validates an INTERVAL literal against an MICROSECOND interval qualifier.
+   *
+   * @throws org.apache.calcite.runtime.CalciteContextException if the interval
+   * value is illegal
+   */
+  private int[] evaluateIntervalLiteralAsMicrosecond(
+      RelDataTypeSystem typeSystem, int sign,
+      String value,
+      String originalValue,
+      SqlParserPos pos) {
+    BigDecimal microsecond;
+
+    // validate as MINUTE(startPrecision), e.g. 'MM'
+    String intervalPattern = "(\\d+)";
+
+    Matcher m = Pattern.compile(intervalPattern).matcher(value);
+    if (m.matches()) {
+      // Break out  field values
+      try {
+        microsecond = parseField(m, 1);
+      } catch (NumberFormatException e) {
+        throw invalidValueException(pos, originalValue);
+      }
+
+      // Validate individual fields
+      checkLeadFieldInRange(typeSystem, sign, microsecond, TimeUnit.MICROSECOND, pos);
+
+      // package values up for return
+      return fillIntervalValueArray(sign, ZERO, microsecond, ZERO);
+    } else {
+      throw invalidValueException(pos, originalValue);
+    }
+  }
+
+  /**
+   * Validates an INTERVAL literal against an NANOSECOND interval qualifier.
+   *
+   * @throws org.apache.calcite.runtime.CalciteContextException if the interval
+   * value is illegal
+   */
+  private int[] evaluateIntervalLiteralAsNanosecond(
+      RelDataTypeSystem typeSystem, int sign,
+      String value,
+      String originalValue,
+      SqlParserPos pos) {
+    BigDecimal nanosecond;
+
+    // validate as MINUTE(startPrecision), e.g. 'MM'
+    String intervalPattern = "(\\d+)";
+
+    Matcher m = Pattern.compile(intervalPattern).matcher(value);
+    if (m.matches()) {
+      // Break out  field values
+      try {
+        nanosecond = parseField(m, 1);
+      } catch (NumberFormatException e) {
+        throw invalidValueException(pos, originalValue);
+      }
+
+      // Validate individual fields
+      checkLeadFieldInRange(typeSystem, sign, nanosecond, TimeUnit.NANOSECOND, pos);
```

Comment:
Ok, the only thing that I see changing between these 3 functions is this TimeUnit. Can you consolidate these three to 1 function, and just pass in the timeunit?

Reply:
Alternately, you can keep these all as separate functions, and just have them all call a helper function.